src/shared/validate/validate.ts

```diff
@@ -0,0 +1,139 @@
+const rfc3986 = new RegExp(uri);
+
+const isValidRFC3986 = (str: any) => {
+  return rfc3986.test(str);
+};
+
+export async function validateW3C<T extends OpenAttestationDocument>(
+  credential: VerifiableCredential<T>,
+  validateTypeWithContext = true
+): Promise<void> {
+  // ensure first context is 'https://www.w3.org/2018/credentials/v1' as it's mandatory, see https://www.w3.org/TR/vc-data-model/#contexts
+  if (
+    !Array.isArray(credential["@context"]) ||
+    (Array.isArray(credential["@context"]) && credential["@context"][0] !== "https://www.w3.org/2018/credentials/v1")
+  ) {
+    throw new Error("https://www.w3.org/2018/credentials/v1 needs to be first in the list of contexts");
+  }
+
+  // ensure issuer is a valid URI according to RFC3986 IF it is a string
+  // TODO check if credential.issuer is string first, as it can be an object containing an id property
+  const issuerId = getId(credential.issuer);
+  if (!isValidRFC3986(issuerId)) {
+    throw new Error("Property 'issuer' id must be a valid RFC 3986 URI");
+  }
+
+  // ensure issuanceDate is a valid RFC3339 date, see https://www.w3.org/TR/vc-data-model/#issuance-date
+  if (!isValidRFC3339(credential.issuanceDate)) {
+    throw new Error("Property 'issuanceDate' must be a valid RFC 3339 date");
+  }
+  // ensure expirationDate is a valid RFC3339 date, see https://www.w3.org/TR/vc-data-model/#expiration
+  if (credential.expirationDate && !isValidRFC3339(credential.expirationDate)) {
+    throw new Error("Property 'expirationDate' must be a valid RFC 3339 date");
+  }
+
+  // https://www.w3.org/TR/vc-data-model/#types
+  if (!credential.type || !Array.isArray(credential.type)) {
+    throw new Error("Property 'type' must exist and be an array");
+  }
+  if (!credential.type.includes("VerifiableCredential")) {
+    throw new Error("Property 'type' must have VerifiableCredential as one of the items");
+  }
+  // if (credential["@context"].length > 1) {
+  //   // e.g. type: ["VerifiableCredential"]
+  //   if (Array.isArray(credential.type) && credential.type.length == 1) {
+  //     throw new Error("Property 'type' must have VerifiableCredential as one of the items");
+  //   }
+  //   // e.g. type: "VerifiableCredential"
+  //   if (typeof credential.type == "string") {
+  //     throw new Error("Property 'type' must be string that is 'VerifiableCredential'");
+  //   }
+  // }
+  // const compacted = await compact(credential, "https://www.w3.org/2018/credentials/v1");
+  // console.log(JSON.stringify(compacted, null, 2));
+  await compact(credential, "https://www.w3.org/2018/credentials/v1", {
+    expansionMap: info => {
+      console.log(info);
+      if (info.unmappedProperty) {
+        throw new Error("The property '" + info.unmappedProperty + "' in the input was not defined in the context");
+      }
+    }
+  });
+
+  // console.log(compacted);
+
+  if (validateTypeWithContext) {
```

Comment:
to remove

Reply:
removed the `validateTypeWithContext` option in d67df54